lib/ohai/plugins/ec2.rb

```diff
@@ -48,25 +65,26 @@ def has_ec2_dmi?
     begin
       # detect a version of '4.2.amazon'
       if dmi[:bios][:all_records][0][:Version] =~ /amazon/
-        Ohai::Log.debug("has_ec2_dmi? == true")
+        Ohai::Log.debug("ec2 plugin: has_ec2_dmi? == true")
         true
       end
     rescue NoMethodError
-      Ohai::Log.debug("has_ec2_dmi? == false")
+      Ohai::Log.debug("ec2 plugin: has_ec2_dmi? == false")
       false
     end
   end
 
-
   def looks_like_ec2?
-    # Try non-blocking connect so we don't "block" if
+    return true if hint?('ec2')
+
+    # if has ec2 mac try non-blocking connect so we don't "block" if
     # the Xen environment is *not* EC2
-    hint?('ec2') || ( has_ec2_dmi? || has_ec2_mac?) && can_metadata_connect?(Ohai::Mixin::Ec2Metadata::EC2_METADATA_ADDR,80)
+    return true if (has_ec2metadata_bin? || has_ec2_dmi?) || (has_xen_mac? && can_metadata_connect?(Ohai::Mixin::Ec2Metadata::EC2_METADATA_ADDR,80))
```

Comment:
The original (pre-8.10) behavior was hint or xen mac + try to connect.  I incorrectly changed it in 8.10 so that it would try to connect if it found the DMI data.  The connection attempt was only in place because the MAC detection triggers on any Xen based system and is very unreliable.  If we find the binary or the DMI data we know 100% that we're on Amazon so there's no need to try to open the socket before we later try to open the socket again.